src/raft/recv_install_snapshot.c

```diff
@@ -478,32 +491,116 @@ static void follower_sent_cb(struct sender *s, int rc)
 	struct rpc *rpc = CONTAINER_OF(s, struct rpc, sender);
 	struct follower *follower = CONTAINER_OF(rpc, struct follower, rpc);
 
+	PRE(follower->ops->is_main_thread());
 	if (UNLIKELY(rc != 0)) {
 		sm_move(&rpc->sm, RPC_ERROR);
 		return;
 	}
-
 	follower_tick(follower, M_MSG_SENT);
 }
 
 static bool is_a_trigger_leader(const struct leader *leader, const struct raft_message *incoming)
 {
-	(void)leader;
-	(void)incoming;
-	return true;
+	int state = sm_state(&leader->sm);
+
+	/* Special cases: */
+	if (incoming == M_WORK_DONE) {
+		return state == LS_HT_WAIT
+		    || state == LS_PAGE_SENT
+		    || state == LS_PERSISTED_SIG_PART
+			|| state == LS_PAGE_READ;
+	} else if (incoming == M_MSG_SENT) {
+		if (sm_state(&leader->rpc.sm) != RPC_FILLED) {
+			return false;
+		}
+		return state == LS_PAGE_READ
+		    || state == LS_SNAP_DONE
+		    || state == LS_F_NEEDS_SNAP
+		    || state == LS_REQ_SIG_LOOP
+		    || state == LS_CHECK_F_HAS_SIGS;
+	} else if (incoming == M_TIMEOUT) {
+		return state == LS_PAGE_READ
+		    || state == LS_SNAP_DONE
+		    || state == LS_F_NEEDS_SNAP
+		    || state == LS_REQ_SIG_LOOP
+		    || state == LS_CHECK_F_HAS_SIGS
+		    || state == LS_WAIT_SIGS;
+	}
+
+	/* From now on, the message pointer is a valid pointer. */
+	PRE(incoming != M_MSG_SENT && incoming != M_TIMEOUT &&
+			incoming != M_WORK_DONE);
```

Comment:
Do we really need it here? 
It's quite clear from the lines above that this assumption is obvious.

Reply:
The only reason is that from this line on this is a valid pointer. If you were to do `msg->type` on `M_MSG_SENT`, or any of the others, the program would crash.